base/show.jl

```diff
@@ -1699,10 +1699,32 @@ function show_unquoted(io::IO, val::SSAValue, ::Int, ::Int)
     if get(io, :maxssaid, typemax(Int))::Int < val.id
         # invalid SSAValue, print this in red for better recognition
         printstyled(io, "%", val.id; color=:red)
+    elseif haskey(io, :compacted) && !io[:compacted]
+        # in the uncompacted part of IncrementalCompact, SSAValues refer to old values
+        printstyled(io, "%", val.id; color=:yellow)
     else
         print(io, "%", val.id)
     end
 end
+
+function show_unquoted(io::IO, val::Core.Compiler.OldSSAValue, ::Int, ::Int)
+    if haskey(io, :compacted) && io[:compacted]
+        # in the compacted part of IncrementalCompact, OldSSAValues refer to old values
+        printstyled(io, "%", val.id; color=:yellow)
+    else
+        print(io, "OldSSAValue(%", val.id, ")")
+    end
+end
+
+function show_unquoted(io::IO, val::Core.Compiler.NewSSAValue, ::Int, ::Int)
+    if haskey(io, :compacted) && !io[:compacted]
```

Comment:
```suggestion
    if !(get(io, :compacted, false)::Bool)
```
Shouldn't the original condition be reverted?

Reply:
I don't think so?

> NewSSAValue in the uncompacted part, which should print regular

Reply:
But from my limited understanding of IncrementalCompact I agree that NewSSAValues in the compacted part may also have to be rendered as regular SSAValues, but I wasn't sure how to convert the negative SSA index to one that makes sense.

Reply:
`NewSSAValues` in the compacted IR should probably be rendered in green, like pending statements in regular IRCode.

Reply:
So what is the meaning of the negative index of a NewSSAValue? An index into `compact.new_new_nodes`? That means they should disappear after compaction, right?

If so, what am I doing wrong here (where I still have a NewSSAValue post-compaction):

```julia
julia> foo(i) = i + 42
       ir = only(Base.code_ircode(foo, (Int,)))[1]
2 1 ─ %1 = Base.add_int(_2, 42)::Int64
3 └──      return %1


julia> compact = Core.Compiler.IncrementalCompact(ir)
──────────────────────────────────────────
2 1 ─ %1 = Base.add_int(_2, 42)::Int64
3 └──      return %1


julia> # partial compaction
       state = Core.Compiler.iterate(compact); compact
2 1 ─ %1 = Base.add_int(_2, 42)::Int64
──────────────────────────────────────────
3 └──      return %1


julia> # insert and replace instruction instruction
       add_stmt = compact.result[1];
julia> inst = Core.Compiler.NewInstruction(Expr(:call, add_stmt[:inst].args[1],
              add_stmt[:inst].args[2], 999), Int) |> Core.Compiler.effect_free
julia> node = Core.Compiler.insert_node!(compact, Core.SSAValue(1), inst)
Core.Compiler.NewSSAValue(-1)
julia> Core.Compiler.setindex!(add_stmt, node, :inst);
julia> compact
2 1 ─      Base.add_int(_2, 999)::Int64
  │   %1 = Core.Compiler.NewSSAValue(-1)::Int64
──────────────────────────────────────────
3 └──      return %1


julia> # finish compaction
       while state !== nothing
           state = Core.Compiler.iterate(compact, state[2])
       end
julia> ir = Core.Compiler.complete(compact)
2 1 ─      Base.add_int(_2, 999)::Int64
  │   %1 = Core.Compiler.NewSSAValue(-1)::Int64
3 └──      return %1


julia> # maybe I should compact again to remove the NewSSAValue?
       ir = Core.Compiler.compact!(ir)
ERROR: BoundsError: attempt to access 3-element Vector{Int64} at index [-1]
Stacktrace:
 [1] getindex
   @ ./essentials.jl:13 [inlined]
 [2] renumber_ssa2
   @ ./compiler/ssair/ir.jl:1037 [inlined]
 [3] renumber_ssa2!(stmt::Any, ssanums::Vector{Any}, used_ssas::Vector{Int64}, new_new_used_ssas::Vector{Int64}, late_fixup::Vector{Int64}, result_idx::Int64, do_rename_ssa::Bool)
   @ Core.Compiler ./compiler/ssair/ir.jl:1061
 [4] process_node!(compact::Core.Compiler.IncrementalCompact, result_idx::Int64, inst::Core.Compiler.Instruction, idx::Int64, processed_idx::Int64, active_bb::Int64, do_rename_ssa::Bool)
   @ Core.Compiler ./compiler/ssair/ir.jl:1238
 [5] iterate_compact(compact::Core.Compiler.IncrementalCompact, ::Tuple{Int64, Int64})
   @ Core.Compiler ./compiler/ssair/ir.jl:1496
 [6] iterate
   @ ./compiler/ssair/ir.jl:1425 [inlined]
 [7] compact!(code::Core.Compiler.IRCode, allow_cfg_transforms::Bool)
   @ Core.Compiler ./compiler/ssair/ir.jl:1685
 [8] compact!(code::Core.Compiler.IRCode)
   @ Core.Compiler ./compiler/ssair/ir.jl:1683
 [9] top-level scope
   @ REPL[22]:1
```